server/src/main/java/org/opensearch/common/util/FastLongHash.java

```diff
@@ -0,0 +1,307 @@
+     */
+    public long add(final long key) {
+        final long ordinal = find(key);
+        if (ordinal != -1) {
+            return -1 - ordinal;
+        }
+
+        if (size >= grow) {
+            grow();
+        }
+
+        return insert(key);
+    }
+
+    /**
+     * Returns the key associated with the given ordinal.
+     * The result is undefined for an unused ordinal.
+     */
+    public long get(final long ordinal) {
+        return keys.get(ordinal);
+    }
+
+    /**
+     * Returns the ordinal associated with the given key, or -1 if the key doesn't exist.
+     *
+     * <p>
+     * Using the 64-bit hash value, up to 32 least significant bits (LSB) are used to identify the
+     * home slot in the hash table, and an additional 16 bits are used to identify the fingerprint.
+     * The fingerprint further increases the entropy and reduces the number of false lookups in the
+     * keys' table during equality checks, which is expensive due to an uncorrelated memory lookup.
+     *
+     * <p>
+     * Total entropy bits = 16 + log2(capacity)
+     *
+     * <p>
+     * Linear probing starts from the home slot, until a match or an empty slot is found.
+     * Values are first checked using their fingerprint (to reduce false positives), then verified
+     * in the keys' table using an equality check.
+     */
+    public long find(final long key) {
+        final long hash = hash(key);
+        final long fingerprint = hash & MASK_FINGERPRINT;
+
+        for (long idx = hash & mask, value, ordinal;; idx = (idx + 1) & mask) {
+            if ((value = table.get(idx)) == -1) {
+                return -1;
+            } else if (((value & MASK_FINGERPRINT) == fingerprint) && (keys.get((ordinal = (value & MASK_ORDINAL))) == key)) {
+                return ordinal;
+            }
+        }
+    }
+
+    /**
+     * Returns the number of mappings in this hash table.
+     */
+    public long size() {
+        return size;
+    }
+
+    /**
+     * Inserts the given key in the hash table and returns its ordinal.
+     *
+     * <p>
+     * Inspired by Robin Hood Hashing (RHH): if the PSL for the existing value is less than the PSL
+     * for the value being inserted, swap the two values and keep going. Values that were inserted
+     * early and thus "lucked out" on their PSLs will gradually be moved away from their preferred
+     * slot as new values come in that could make better use of that place in the table. It evens out
+     * the PSLs across the board and reduces the longest PSL dramatically.
+     *
+     * <p>
+     * A lower variance is better because, with modern CPU architectures, a PSL of 1 isn't much
+     * faster than a PSL of 3, because the main cost is fetching the cache line. The ideal hash
+     * table layout is the one where all values have equal PSLs, and that PSL fits within one cache line.
+     *
+     * <p>
+     * The expected longest PSL for a full table: <code>log(N)</code>
```

Comment:
What would be the longest PSL in the existing approach? is it N?

Reply:
That's right. Expected longest PSL for a full table in the existing approach would be Θ(N).